lib/vm/src/instance/builder.rs

```diff
@@ -0,0 +1,197 @@
+use super::{Instance, InstanceAllocator};
+use crate::vmcontext::{VMMemoryDefinition, VMTableDefinition};
+use crate::{ModuleInfo, VMOffsets};
+use std::alloc::{self, Layout};
+use std::convert::TryFrom;
+use std::mem;
+use std::ptr::{self, NonNull};
+use wasmer_types::entity::EntityRef;
+use wasmer_types::{LocalMemoryIndex, LocalTableIndex};
+
+/// This is an intermediate type that manages the raw allocation and
+/// metadata when creating an [`Instance`].
+///
+/// This type will free the allocated memory if it's dropped before
+/// being used.
+///
+/// The [`InstanceBuilder::instance_layout`] computes the correct
+/// layout to represent the wanted [`Instance`].
+///
+/// Then we use this layout to allocate an empty `Instance` properly.
+pub struct InstanceBuilder {
+    /// The buffer that will contain the [`Instance`] and dynamic fields.
+    instance_ptr: NonNull<Instance>,
+    /// The layout of the `instance_ptr` buffer.
+    instance_layout: Layout,
+    /// Information about the offsets into the `instance_ptr` buffer for
+    /// the dynamic fields.
+    offsets: VMOffsets,
+    /// Whether or not this type has transferred ownership of the
+    /// `instance_ptr` buffer. If it has not when being dropped,
+    /// the buffer should be freed.
+    consumed: bool,
+}
+
+impl Drop for InstanceBuilder {
+    fn drop(&mut self) {
+        if !self.consumed {
+            // If `consumed` has not been set, then we still have ownership
+            // over the buffer and must free it.
+            let instance_ptr = self.instance_ptr.as_ptr();
+            unsafe {
+                std::alloc::dealloc(instance_ptr as *mut u8, self.instance_layout);
+            }
+        }
+    }
+}
+
+impl InstanceBuilder {
+    /// Allocates instance data for use with [`InstanceHandle::new`].
+    ///
+    /// Returns a wrapper type around the allocation and 2 vectors of
+    /// pointers into the allocated buffer. These lists of pointers
+    /// correspond to the location in memory for the local memories and
+    /// tables respectively. These pointers should be written to before
+    /// calling [`InstanceHandle::new`].
+    pub fn new(
+        module: &ModuleInfo,
+    ) -> (
+        InstanceBuilder,
+        Vec<NonNull<VMMemoryDefinition>>,
+        Vec<NonNull<VMTableDefinition>>,
+    ) {
+        let offsets = VMOffsets::new(mem::size_of::<usize>() as u8, module);
+        let instance_layout = Self::instance_layout(&offsets);
+
+        #[allow(clippy::cast_ptr_alignment)]
+        let instance_ptr = unsafe { alloc::alloc(instance_layout) as *mut Instance };
+
+        let instance_ptr = if let Some(ptr) = NonNull::new(instance_ptr) {
+            ptr
+        } else {
+            alloc::handle_alloc_error(instance_layout);
+        };
+
+        let unprepared = Self {
```

Comment:
`let builder`? :-)

Reply:
We are going with allocator. So perhaps `let allocator`?